### PR TITLE
Adding STM32 support and libraries to monorepo

### PR DESCRIPTION
**Pull request by ayushchakra**
This PR adds support for building and flashing binary files for the STM32f1xx microcontroller family. As a proof of concept, `examples/blinky_arm` was created to demonstrate a LED blinking on the STM32F103RBT6.

Outstanding Tasks (For a follow up PR)

- [ ] Figure out how to submodule `libopencm3` so that the headers don't need to be copied (started in `ayush/stm32_submodule_backup`
- [ ] Figure out how to maintain one `cc_firmware` function, rather than what is done in this PR, which is creating a new `cc_arm_firmware` wrapper to build firmware.
- [ ] Update `libs/*` to support stm32s.

### CODE REVIEW

File: third_party/libopencm3/BUILD [hunk 1]
```diff
@@ -0,0 +1,8 @@
+cc_library(
+    name = "libopencm3",
+    hdrs = glob(["include/**"]),
```
Comment (jack-greenberg) [inline third_party/libopencm3/BUILD:3]: need srcs too, right?

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
+    "files.associations": {
+        "common.h": "c"
```
Comment (ayushchakra) [inline .vscode/settings.json:3]: don't push this

File: WORKSPACE [hunk 3]
```diff
@@ -19,6 +20,58 @@ http_archive(
...
+)
+
+load("@arm_none_eabi//:deps.bzl", "arm_none_eabi_deps")
+
+arm_none_eabi_deps()
+
+# Libopencm3 (HAL for STM32)
+http_archive(
+    name = "libopencm3",
+    build_file = "@//third_party/libopencm3:BUILD",
```
Comment (ayushchakra) [inline WORKSPACE:38]: clean this file up to be up to date with what is actually used

File: bazel/defs.bzl [hunk 4]
```diff
@@ -375,17 +421,99 @@ def cc_firmware(name, **kwargs):
...
+        copts = kwargs.pop("copts")
+
+    linkopts = []
+    if kwargs.get("linkopts"):
+        linkopts = kwargs.pop("linkopts")
+
+    cc_binary(
+        name = "{}.elf".format(name),
+        linkopts = linkopts + select({
+            "//bazel/constraints:stm32f103rbt6": ["-T $(location //scripts/ldscripts:stm32f103rbt6.ld)", "-Os", "-std=c99", "-ggdb3", "-mcpu=cortex-m3", "-mthumb", "-msoft-float", "-fno-common", "-ffunction-sections", "-fdata-sections", "-Wextra", "-Wshadow", "-Wno-unused-variable", "-Wimplicit-function-declaration", "-Wredundant-decls", "-Wstrict-prototypes", "-Wmissing-prototypes", "-MD", "-Wall", "-Wundef","-nostartfiles", "-mcpu=cortex-m3", "-mthumb", "-msoft-float", "-specs=nano.specs", "-Wl,--gc-sections", "-Wl,--start-group", "-lc", "-lgcc", "-lnosys", "-Wl,--end-group"],
```
Comment (ayushchakra) [inline bazel/defs.bzl:458]: move this

File: test_stm_image/BUILD [hunk 5]
```diff
@@ -0,0 +1,15 @@
+load("//bazel:defs.bzl", "cc_arm_firmware")
+
+cc_arm_firmware(
+    name = "test_stm_image",
+    srcs = [
```
Comment (ayushchakra) [inline test_stm_image/BUILD:5]: TODO: integrate into examples/blinky

**Review by Rixhee — APPROVED** [review]

**Pull request merged**